File: core/src/main/java/net/year4000/mapnodes/regions/CuboidRegion.java
/*
 * Copyright 2016 Year4000. All Rights Reserved.
 */

package net.year4000.mapnodes.regions;

import com.flowpowered.math.vector.Vector3i;
import com.google.common.collect.ImmutableSet;
import net.year4000.utilities.Conditions;
import net.year4000.utilities.Utils;

/** Represents a cuboid region with two positions */
public final class CuboidRegion extends AbstractComplexRegion {
    private Vector3i pointOne;
    private Vector3i pointTwo;

    public CuboidRegion(Vector3i pointOne, Vector3i pointTwo) {
        this.pointOne = Conditions.nonNull(pointOne, "pointOne");
        this.pointTwo = Conditions.nonNull(pointTwo, "pointTwo");
    }

    /** Generate the points of the cuboid region */
    @Override
    protected ImmutableSet<Vector3i> generatePoints() {
        ImmutableSet.Builder<Vector3i> points = ImmutableSet.builder();
        Vector3i min = pointOne.min(pointTwo);
        Vector3i max = pointOne.max(pointTwo);
        for (int y = min.getY(); y < max.getY(); y++) {
            for (int x = min.getX(); x < max.getX(); x++) {
                for (int z = min.getZ(); z < max.getZ(); z++) {
                    points.add(new Vector3i(x, y, z));
                }
            }
        }
        return points.build();
    }

    @Override
    public boolean contains(Vector3i vector3i) {
        Vector3i min = pointOne.min(pointTwo);
        Vector3i max = pointOne.max(pointTwo);
        boolean x = vector3i.getX() >= min.getX() && vector3i.getX() <= max.getX();
        boolean y = vector3i.getY() >= min.getY() && vector3i.getY() <= max.getY();
        boolean z = vector3i.getZ() >= min.getZ() && vector3i.getZ() <= max.getZ();
        return x && y && z;
    }

    @Override
    public boolean equals(Object other) {
        return Utils.equals(this, other);
    }

    @Override
    public int hashCode() {
        return Utils.hashCode(this);
    }

    @Override
    public String toString() {
        return Utils.toString(this);
    }
}

File: core/src/main/js/net/year4000/mapnodes/js/regions/cuboid_region.js
/*
 * Copyright 2017 Year4000. All Rights Reserved.
 */

/** Represents a cuboid region with two positions */
class CuboidRegion extends AbstractRegion {

  /** point_one and point_two are both vector3 */
  constructor(point_one, point_two) {
    this._point_one = Conditions.not_null(point_one, 'point_one')
    this._point_two = Conditions.not_null(point_two, 'point_two')
  }

  /** Checks if the vector contains in this cuboid */
  contains(vector3) {
    let min = this._point_one.clone().min(this._point_two)
    let max = this._point_one.clone().max(this._point_two)
    let x = vector3.x >= min.x && vector3.x <= max.x
    let y = vector3.y >= min.y && vector3.y <= max.y
    let z = vector3.z >= min.z && vector3.z <= max.z
    return x && y && z
  }

  /** Generate all the points in this cuboid */
  get _points() {
    let points = new Immutable.Set.of();
    let min = this._point_one.clone().min(this._point_two)
    let max = this._point_one.clone().max(this._point_two)
    for (let y = min.y; y < max.y; y++) {
      for (let x = min.x; x < max.x; x++) {
        for (let z = min.z; z < max.z; z++) {
          points = points.add(new THREE.Vector3(x, y, z));
        }
      }
    }
    return points;
  }

  /** Get all the points this cuboid has */
  get points() {
    return _.memoize(this._points);
  }

  /** Checks if the two cuboids are equal */
  equals(other) {
    return other && _.isEqual(this._point_one, other._point_one) && _.isEqual(this._point_two, other._point_two);
  }
}
